src/clj/fluree/db/reasoner.cljc

```diff
@@ -218,24 +235,42 @@
             (recur r db**))
           db*)))))
 
+(defn deduplicate-raw-rules
+  [raw-rules]
+  (let [rule-ids (map first raw-rules)
+        duplicate-ids (filter #(< 1 (last %)) (frequencies rule-ids))]
+    (reduce (fn [rules [duplicate-id occurances]]
```

Comment:
It looks like this attempts to retain every rule that has identical names by modifying their names with an indexed value.

I guess the options here would be:
1) Throw an exception
2) "merge" behavior, where only one is retained (with the idea they should be identical if they have the identical IRI)
3) renaming duplicated names

You opted for #3 - is there a reason for that over the others?

Reply:
My thinking for option 3 was that if we are getting two rules from different sources with the same IRI, we would still want to preserve the functionality of both of those rules. If we wanted to retain one and remove the other, we would probably want the person submitting the query to have control over that which gets complicated I think. We also discussed this a few weeks ago and that was the solution that you leaned towards at the time.

Reply:
Apologies if I leaned towards that previously, sometimes it takes me getting into things deeper to have a more natural feel of what seems right.

I think I'd lean towards (2) now, and use merge semantics (left to right). This would allow the flexibility of overwriting rules. I could see this if global governance had a rule set that must be followed, regardless of if a local db owner tried to overwrite the rule.

While the rules still remain an 'or' condition, once we had a rule designated as "important" (similar to CSS's important!) it would allow and conditions.

The combination of merge semantics + important would allow global governance to guarantee a certain set of rules are always followed. Without it, I think there could be possibilities that folks could get around rules.

Also, with the suggestion below of:
```
(-> db 
    (reason :owl [{...some json-ld-rules}])
    (reason :datalog some-rules-db)
    (reason :owl some-other-rules-db))
```

It allows the flexibility of holding a db, applying some query/tx on it, then applying another rule set and applying another rule set. Supporting a list here likely precludes changing rules on an existing db, it would have to be a "reset"... whereas adding rules one at a time allows more flexibility.

Reply:
In option you say that we should be doing merge behavior under the assumption that identical ids should have identical rules, but if we're making that assumption, then we wouldn't need to make any changes here because one of the rules already gets overridden in that case. This is moreso meant to solve the problem of identical ids with distinct rules. In that case, should we be throwing an exception?

I also think I'm not fully understanding how an `important` tag enables `and` conditions. If there are two rules applied as an `or` condition, then for a given subject, it will be returned if either one of those rules matches it. For an `and` condition, a subject would only be returned if both rules match it, but if we have two rules and one of them is getting overridden because the other rule is marked as `important`, then it's not evaluating the other rule at any point.

And what should we do if neither rule is marked as `important` or both rules are marked as `important`?

If something is marked as `important` I don't think we need merge semantics to ensure that those are always followed. I would just have to add a check in the deduplicate function that if any of the rules with duplicate IRIs are marked as `important` then we ignore the rest of those duplicated rules.

If we are adding rules one at a time, I think that might cause issues where the order that we apply the rules becomes much more important. If we have a rule that adds `ex:uncle` to an entity under some condition and another rule that adds `ex:aunt` to an entity if they are married to an `ex:uncle`, then the order that those rules are applied will determine whether or not `ex:aunt` gets added to that entity.

Reply:
Merge behavior with two identical rule ids could:
a) overwrite without warning
b) overwrite with warning
c) throw

To me (b) feels like the right answer, not that it needs to be this way and do everything in a single PR.

I think we should strive for predictable, simple behavior and merge feels right (perhaps not perfect). I think it can provide a functional feature as well, which is if there were a common set of rules in some ledger, I could "deactivate/replace" some of them by merging another rule set with just those ids.

As for `important` (or a different semantic might be better), the idea is to allow AND conditions. Right now only OR conditions are supported. By marking a rule as `important` it means it *must* pass. If there are 2+ `important` rules for the same Flake, they must all pass (AND). In that case none of the other rules matter which are not `important` and they can be effectively ignored. Even though it might be ignored for a specific Flake, it might be relevant for other flakes as rules can have narrow or broad scopes.

Again, not all for the same PR - just explaining the breadth of what I think the reasoning feature can do, as it helps inform the right behavior of what to do when we have multiple reasoning rule sets which is the focus here.

Reply:
How do you warn? If it's just log messages most users will never see them. I would lean towards (a) with documentation or (c), I think.

Reply:
I think with (c) you'd loose a valuable feature. I think (a) is definitely an option. I still prefer (b) with log messages.

Your assumption that they won't see the log messages assumes this is coming from someone running a production server. I think folks who are developing and merging reasoning rules are likely be running this as a developer locally getting the behavior they are looking for. If for personal analytics, it will never run as a production server - or if a production environment that logic they develop locally will be part of their app. If they want to turn off the logging once in production they can easily do so with logback.xml file.

Reply:
Ok, I misunderstood your comment about the `important` tag to be saying that that's how you think conflicting rule ids should be resolved. You're just saying that you think the merge behavior makes more sense with the eventual implementation of the `important` tag. I have more questions about the implementation of that, but it can wait until we are actually thinking of working on it.

So for this PR then, how do we know which source of rules to prioritize for overriding then? You mentioned in one of your other comments that the rule sources would be better as just a list instead of using the `:rule-graphs` and `:rule-dbs` map syntax, are you thinking it just takes that list and merges from left to right regardless of what type of source it is coming from?

Reply:
I think it will require keeping a map of rules on the db with ids as keys (or similar) - for the primary purpose of merging in new rules if added and then applying the result of that merge as the working rule set.

Reply:
Maybe I'm not understanding, but are you saying that the ids of the rules in the provided rule sources should be transacted to the db? I'm not sure what that accomplishes over just getting the raw rules in `reasoner/reason` and then merging those maps together?

Reply:
No, they shouldn't be transacted - but they need to be retained somewhere... e.g. `(:reasoning-rules db)`... so we are hanging them on the db object as a new key-val pair. This way if more rules are added later we can perform the merge and apply the result as the new set of rules.

The other thing that needs to be considered is equality checks on the dbs that are reasoned, but that could be addressed separately.